### DISCUSSION TITLE
How to use termwiz with std::process::Child

### DISCUSSION THREAD

**Discussion by melody-rs** [wezterm/wezterm#3376] [Q&A] [answered]
I'm writing an egui app that needs to spawn processes and I'd ideally like to capture their output and display it. termwiz seems like a good fit since it can handle a lot of logic I'd much rather not have to write myself, and probably functions better than anything I can write.

Anyway, my code is like this:
```rs
/* in application code */
if ui.button("Run").clicked() {
    match std::process::Command::new(&self.command_name)
        .stdout(std::process::Stdio::piped())
        .stdin(std::process::Stdio::piped())
        .spawn()
    {
        Ok(child) => match Terminal::new(child) {
            Ok(t) => self.terminals.push(t),
            Err(e) => eprintln!("error creating terminal: {e:?}"),
        },
        Err(e) => eprintln!("error starting process: {e:?}"),
    }
}

/* elsewhere */
pub struct Terminal {
    terminal: BufferedTerminal<SystemTerminal>,
}

impl Terminal {
    pub fn new(mut child: std::process::Child) -> Result<Self, termwiz::Error> {
        let stdout = child
            .stdout
            .take()
            .ok_or_else(|| "no stdout on child".to_string())?;
        let stdin = child
            .stdin
            .take()
            .ok_or_else(|| "no stdout on child".to_string())?;
        let terminal = SystemTerminal::new_with(Capabilities::new_from_env()?, stdout, stdin)?;
        let terminal = BufferedTerminal::new(terminal)?;

        Ok(Self { terminal })
    }
}
```
No matter what I change, (i.e. dropping the usage of `std::process::Stdio::piped()` and changing `SystemTerminal` to use `new_from_stdio`) I always get this error: 
`Error(StringErr(StringWrap("stdin and stdout must both be tty handles")))`
What am I doing wrong?

**Comment by wez** [comment, marked as answer]
The wezterm repo has crates that can potentially help you at different layers:

* portable_pty - allows you to spawn a process into a PTY and control the screen size of the PTY and read/write its output/input streams. There is no terminal emulation performed by this layer, but it will let you work with PTYs on all major operating systems.
* termwiz:
    *  functions that can parse terminal escape sequences so that you can reason about them, for both input (keys/mouse) and output
    *  functions that allow you to compose output for building TUI applications
    * The `Terminal` type in termwiz is geared up for authoring TUI applications that run in an existing terminal
* wezterm-term:
    * Implements a terminal emulator that understands escape sequences and builds an internal model of what the terminal screen and scrollback would look like.
    * It builds on top of the types in the termwiz crate
    * The `Terminal` type in wezterm-term is a terminal emulator that can be embedded into another application that wants to take responsibility for rendering what that might look like
* wezterm / wezterm-gui
   * Implements a multiplexing terminal emulator GUI, supporting multiple panes, tabs, windows, remote connectivity and more

I'm not super clear what you're trying to do; can you share a bit more about your goals?

**Reply by melody-rs** [reply]
I'm making a game engine editor called [Luminol](https://github.com/Astrabit-ST/Luminol) which is a remake of a pretty old piece of software, called RPG Maker XP

RPG Maker has the ability to playtest your game right from the editor, I'd like to expand upon that functionality tho by embedding a terminal inside Luminol to monitor the stdout/stdin of your game.

Based on what you said, it sounds like wezterm-term would be what I should be using?

**Reply by wez** [reply]
It depends on the complexity of the game output. If it only uses simple escape sequences, it may be simple enough to parse and handle the screen updates yourself.

wezterm-term isn't considered to have a totally stable public API (being intended primarily for wezterm) and the crate isn't published on crates.io as a result.  You can embed it by referencing the git repo from your cargo.toml; make sure you pin to a specific commit so that you are not taken by surprise if/when something changes.

wezterm-term will save you from having to build the complex terminal emulation part.

You still need to retrieve the `Line`s and render the `Cell`s on your game display surface; the cell describes the character attributes of the cells. There are methods for clustering the cells with the attributes so that you can render runs of text with the same attributes.

There are many attributes, but for a basic display you can focus on handling just bold and colors.

**Reply by melody-rs** [reply]
> It depends on the complexity of the game output. If it only uses simple escape sequences, it may be simple enough to parse and handle the screen updates yourself.

True, but I have no idea how to do that ~~and it's very likely someone will be embedding a repl like irb in the output~~

Alright, I'll try and do that then!

**Reply by wez** [reply]
to clarify this:

> parse and handle the screen updates yourself.

I meant that you can parse using the termwiz parser, but handle the painting yourself.

**Reply by melody-rs** [reply]
Ah alright

**Reply by melody-rs** [reply]
I'm assuming I need to impl TerminalConfiguration myself?

**Reply by melody-rs** [reply]
Actually, do you have a discord or something so we can talk better?

**Reply by wez** [reply]
https://app.element.io/#/room/#wezterm:matrix.org
